README.md

## radio_grouped_buttons

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

This Widget is specially designed for List of grouped radio buttons in grid manner flutter.
You can customize it any time with its all attributes. You can create radio button group in horizontal
manner or vertical manner as you wish. Button width,height spaces between buttons,colors all the things
are customizable.


## Installing
```
dependencies:
    radio_grouped_buttons: 1.0.1
```

## Creating radio button
Make sure You have included the parent widget with bounded like container with width and height.
By Default the initial selected value is the first value of the list that you provide for the button
value list.But You can customize your default button value by using initialSelection property.

### All Attributes
```
CustomRadioButton({this.buttonLables,
    this.buttonValues,
    this.radioButtonValue,
    this.buttonWidth,
    this.buttonColor,
    this.selectedColor,
    this.buttonHeight,
    this.horizontal,
    this.enableShape,
    this.elevation,
    this.customShape,
    this.fontSize,
    this.lineSpace,
    this.buttonSpace,
    this.buttonBorderColor,
    this.textColor,
    this.selectedTextColor,
    this.initialSelection,
    this.unselectedButtonBorderColor
  })
```

### creating horizontal radio button list
When It overflow container width It will be added to next line automatically.

```
            Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 160,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                                      print("Button value "+value.toString());
                                      print("Integer value "+index.toString());
                                    },
                  horizontal: true,
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                ),
              ),
```

### creating vertical radio button list
When It overflow container width in one line It will be automatically added to new column
inside the parent widget.

```
            Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                                      print("Button value "+value.toString());
                                      print("Integer value "+index.toString());
                                    },
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                ),
              ),
```

![horizontal and vertical radio buttons](/screenshots/gif2.gif)

### Creating Equal width vertical radio button list

If you want to create the buttons with equal width and height buttonWidth and buttonHeight attributes
are also available

```
            Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                                      print("Button value "+value.toString());
                                      print("Integer value "+index.toString());
                                    },
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  buttonWidth: 150,
                ),
              ),
```

### creating equal width horizontal button list

```
                  Container(
                    padding: EdgeInsets.all(10),
                    width: MediaQuery.of(context).size.width,
                    height: 250,
                    child: CustomRadioButton(
                      buttonLables: buttonList,
                      buttonValues: buttonList,
                      radioButtonValue: (value,index){
                                          print("Button value "+value.toString());
                                          print("Integer value "+index.toString());
                                        },
                      horizontal: true,
                      enableShape: true,
                      buttonSpace: 5,
                      buttonColor: Colors.white,
                      selectedColor: Colors.cyan,
                      buttonWidth: 150,
                    ),
                  )
```

![horizontal and vertical radio buttons](/screenshots/gif1.gif)

### Set The Initial selected value as the Third Button
We can set the initial selected button for any index.But when we are concerning an array we have to
set the initial selected value= index(our button index)-1
For an example If we need to set the initial selected button for the third button in our button list.
when we are concerning an array so our button value is 2.

```
    Container(
                    padding: EdgeInsets.all(10),
                    width: MediaQuery.of(context).size.width,
                    height: 250,
                    child: CustomRadioButton(
                      buttonLables: buttonList,
                      buttonValues: buttonList,
                      radioButtonValue: (value,index){
                        print("Button value "+value.toString());
                        print("Integer value "+index.toString());
                      },
                      horizontal: true,
                      enableShape: true,
                      buttonSpace: 5,
                      initialSelection: 2,
                      buttonColor: Colors.white,
                      selectedColor: Colors.cyan,
                      buttonWidth: 150,
                    ),
                  )
```

CHANGELOG.md
## [0.0.1] - 30/05/2020

* Initial Release.
This is a smart flutter widget that is fully customizable for radio button group creations.

  Features:

     * Equal width buttons
     * radio buttons horizontal
     * radio buttons vertical
     * radio buttons equal width and height

## [1.0.0] - 30/05/2020
## [1.0.0+1]- 1/07/2020
* Readme updated and some issues fixed.
## [1.0.0+2]- 2/08/2020
* Added the default value Configuration.
* Added Unselected Button Bordercolor attribute

## [1.0.1]- 17/10/2021
* Added null safety

lib/custom_buttons/custom_radio_buttons_group.dart
import 'package:flutter/material.dart';

class CustomRadioButton extends StatefulWidget {
  CustomRadioButton({
    this.buttonLables,
    this.buttonValues,
    this.radioButtonValue,
    this.buttonWidth,
    this.buttonColor,
    this.selectedColor,
    this.buttonHeight = 35,
    this.horizontal = false,
    this.enableShape = false,
    this.elevation = 10,
    this.customShape,
    this.fontSize = 15,
    this.lineSpace = 5,
    this.buttonSpace = 5,
    this.buttonBorderColor = const Color(0xFF2594B22B),
    this.unselectedButtonBorderColor = const Color(0xFF2594B22B),
    this.textColor = Colors.black,
    this.selectedTextColor = Colors.white,
    this.initialSelection=0,
  })  : assert(buttonLables!.length == buttonValues!.length),
        assert(buttonLables!.length > initialSelection),
        assert(buttonColor != null),
        assert(selectedColor != null);

  final bool horizontal;
  final double fontSize;
  final List? buttonValues;
  final double buttonSpace;
  final double buttonHeight;

  final double lineSpace;
  final List<String>? buttonLables;

  final Function(dynamic, int)? radioButtonValue;

  final Color? selectedColor;
  final Color buttonBorderColor;
  final Color unselectedButtonBorderColor;
  final Color? buttonColor;
  final Color textColor;
  final Color selectedTextColor;
  final ShapeBorder? customShape;
  final bool enableShape;
  final double elevation;
  final double? buttonWidth;
  final int initialSelection;

  _CustomRadioButtonState createState() => _CustomRadioButtonState();
}

class _CustomRadioButtonState extends State<CustomRadioButton> {
  int currentSelected = 0;
  String? currentSelectedLabel;

  @override
  void initState() {
    super.initState();
    if (widget.initialSelection != null) {
      currentSelectedLabel = widget.buttonLables![widget.initialSelection];
    } else {
      currentSelectedLabel = widget.buttonLables![0];
    }
  }

  List<Widget> buildButtonsColumn() {
    List<Widget> buttons = [];
    for (int index = 0; index < widget.buttonLables!.length; index++) {
      var button = Card(
        color: currentSelectedLabel == widget.buttonLables![index] ? widget.selectedColor : widget.buttonColor,
        elevation: widget.elevation,
        shape: widget.enableShape
            ? widget.customShape == null
                ? RoundedRectangleBorder(
                    borderRadius: BorderRadius.all(Radius.circular(50)),
                  )
                : widget.customShape
            : null,
        child: Container(
          height: widget.buttonHeight,
          width: widget.buttonWidth ?? null,
          constraints: BoxConstraints(maxWidth: 200),
          child: MaterialButton(
            shape: widget.enableShape
                ? widget.customShape == null
                    ? OutlineInputBorder(
                        borderSide: BorderSide(color: currentSelectedLabel == widget.buttonLables![index] ? widget.buttonBorderColor : widget.unselectedButtonBorderColor /*.withOpacity(0.1)*/, width: 1),
                        borderRadius: BorderRadius.all(Radius.circular(20)),
                      )
                    : widget.customShape
                : OutlineInputBorder(
                    borderSide: BorderSide(color: currentSelectedLabel == widget.buttonLables![index] ? widget.buttonBorderColor : widget.unselectedButtonBorderColor /*.withOpacity(0.1)*/, width: 1),
                    borderRadius: BorderRadius.zero,
                  ),
            onPressed: () {
              widget.radioButtonValue!(widget.buttonValues![index], index);
              setState(() {
                currentSelected = index;
                currentSelectedLabel = widget.buttonLables![index];
              });
            },
            child: Text(
              widget.buttonLables![index],
              style: TextStyle(
                color: currentSelectedLabel == widget.buttonLables![index] ? widget.selectedTextColor : widget.textColor,
                fontSize: widget.fontSize,
              ),
            ),
          ),
        ),
      );
      buttons.add(button);
    }
    return buttons;
  }

  List<Widget> buildButtonsRow() {
    List<Widget> buttons = [];
    for (int index = 0; index < widget.buttonLables!.length; index++) {
      var button = Card(
        color: currentSelectedLabel == widget.buttonLables![index] ? widget.selectedColor : widget.buttonColor,
        elevation: widget.elevation,
        shape: widget.enableShape
            ? widget.customShape == null
                ? RoundedRectangleBorder(
                    borderRadius: BorderRadius.all(Radius.circular(50)),
                  )
                : widget.customShape
            : null,
        child: Container(
          height: widget.buttonHeight,
          width: widget.buttonWidth ?? null,
          constraints: BoxConstraints(maxWidth: 200),
          child: MaterialButton(
            shape: widget.enableShape
                ? widget.customShape == null
                    ? OutlineInputBorder(
                        borderSide: BorderSide(color: currentSelectedLabel == widget.buttonLables![index] ? widget.buttonBorderColor : widget.unselectedButtonBorderColor, width: 1),
                        borderRadius: BorderRadius.all(Radius.circular(20)),
                      )
                    : widget.customShape
                : OutlineInputBorder(
                    borderSide: BorderSide(color: currentSelectedLabel == widget.buttonLables![index] ? widget.buttonBorderColor : widget.unselectedButtonBorderColor, width: 1),
                    borderRadius: BorderRadius.zero,
                  ),
            onPressed: () {
              widget.radioButtonValue!(widget.buttonValues![index], index);
              setState(() {
                currentSelected = index;
                currentSelectedLabel = widget.buttonLables![index];
              });
            },
            child: Text(
              widget.buttonLables![index],
              style: TextStyle(
                color: currentSelectedLabel == widget.buttonLables![index] ? widget.selectedTextColor : widget.textColor,
                fontSize: widget.fontSize,
              ),
            ),
          ),
        ),
      );
      buttons.add(button);
    }
    return buttons;
  }

  @override
  Widget build(BuildContext context) {
    return widget.horizontal
        ? Wrap(
            spacing: widget.buttonSpace, // gap between adjacent chips
            runSpacing: widget.lineSpace,
            children: buildButtonsColumn(),
          )
        : Wrap(
            direction: Axis.vertical,
            spacing: widget.buttonSpace, // gap between adjacent chips
            runSpacing: widget.lineSpace,
            children: buildButtonsRow(),
          );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:radio_grouped_buttons/radio_grouped_buttons.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Group Grid View radio Buttons',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: GroupGridRadioButtonExample(),
    );
  }
}

class GroupGridRadioButtonExample extends StatefulWidget {
  @override
  _GroupGridRadioButtonExampleState createState() => _GroupGridRadioButtonExampleState();
}

class _GroupGridRadioButtonExampleState extends State<GroupGridRadioButtonExample> {

  ///horizontal list of buttons
  List<String> buttonList=[
      "Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"];

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        appBar: AppBar(
          title: Text("Group grid Radio Button Example"),
        ),
        body: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              SizedBox(height: 20,),
              Text("Horizontal radio list"),
              ///use this widget in bounded parent widget
              Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 160,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                    print("Button value "+value.toString());
                    print("Integer value "+index.toString());
                  },
                  horizontal: true,
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  //buttonWidth: 150,
                ),
              ),
              SizedBox(height: 20,),


              Text("Vertical radio list"),
              ///use this widget in bounded parent widget
              Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                    print("Button value "+value.toString());
                    print("Integer value "+index.toString());
                  },
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  //buttonWidth: 150,
                ),
              ),
              SizedBox(height: 20,),


              Text("Same length button radio vertical list"),
              ///use this widget in bounded parent widget
              Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                    print("Button value "+value.toString());
                    print("Integer value "+index.toString());
                  },
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  buttonWidth: 150,
                ),
              ),
              SizedBox(height: 20,),


              Text("Same length button radio horizontal list"),
              ///use this widget in bounded parent widget
              Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                    print("Button value "+value.toString());
                    print("Integer value "+index.toString());
                  },
                  horizontal: true,
                  enableShape: true,
                  buttonSpace: 5,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  buttonWidth: 150,
                ),
              ),

              //Initial Value selection
              Text("Set the Initial selected value as the third button"),
              ///use this widget in bounded parent widget
              Container(
                padding: EdgeInsets.all(10),
                width: MediaQuery.of(context).size.width,
                height: 250,
                child: CustomRadioButton(
                  buttonLables: buttonList,
                  buttonValues: buttonList,
                  radioButtonValue: (value,index){
                    print("Button value "+value.toString());
                    print("Integer value "+index.toString());
                  },
                  horizontal: true,
                  enableShape: true,
                  buttonSpace: 5,
                  initialSelection: 2,
                  buttonColor: Colors.white,
                  selectedColor: Colors.cyan,
                  buttonWidth: 150,
                ),
              )

            ],
          ),
        ),
      ),
    );
  }
}

